pkg/detectors/aws/session_keys/sessionkey.go

```diff
@@ -0,0 +1,306 @@
+
+	"github.com/trufflesecurity/trufflehog/v3/pkg/common"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/detectors"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/detectors/aws"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/pb/detectorspb"
+)
+
+type scanner struct {
+	*detectors.CustomMultiPartCredentialProvider
+	verificationClient *http.Client
+	skipIDs            map[string]struct{}
+}
+
+func New(opts ...func(*scanner)) *scanner {
+	scanner := &scanner{
+		skipIDs: map[string]struct{}{},
+	}
+	for _, opt := range opts {
+		opt(scanner)
+	}
+
+	scanner.CustomMultiPartCredentialProvider = detectors.NewCustomMultiPartCredentialProvider(2048) // ????
+	return scanner
+}
+
+func WithSkipIDs(skipIDs []string) func(*scanner) {
+	return func(s *scanner) {
+		ids := map[string]struct{}{}
+		for _, id := range skipIDs {
+			ids[id] = struct{}{}
+		}
+
+		s.skipIDs = ids
+	}
+}
+
+// Ensure the scanner satisfies the interface at compile time.
+var _ interface {
+	detectors.Detector
+	detectors.CustomResultsCleaner
+} = (*scanner)(nil)
+
+var (
+	defaultVerificationClient = common.SaneHttpClient()
+
+	// Key types are from this list https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_identifiers.html#identifiers-unique-ids
+	idPat      = regexp.MustCompile(`\b((?:ASIA)[A-Z0-9]{16})\b`)
+	sessionPat = regexp.MustCompile(`(?:[^A-Za-z0-9+/]|\A)([a-zA-Z0-9+/]{100,}={0,3})(?:[^A-Za-z0-9+/=]|\z)`)
+)
+
+// Keywords are used for efficiently pre-filtering chunks.
+// Use identifiers in the secret preferably, or the provider name.
+func (s scanner) Keywords() []string {
+	return []string{"ASIA"}
+}
+
+// FromData will find and optionally verify AWS secrets in a given set of bytes.
+func (s scanner) FromData(ctx context.Context, verify bool, data []byte) (results []detectors.Result, err error) {
+	logger := logContext.AddLogger(ctx).Logger().WithName("awssessionkey")
+	dataStr := string(data)
+	dataStr = aws.UrlEncodedReplacer.Replace(dataStr)
+
+	// Filter & deduplicate matches.
+	idMatches := make(map[string]struct{})
+	for _, matches := range idPat.FindAllStringSubmatch(dataStr, -1) {
+		idMatches[matches[1]] = struct{}{}
+	}
+	secretMatches := make(map[string]struct{})
+	for _, matches := range aws.SecretPat.FindAllStringSubmatch(dataStr, -1) {
+		secretMatches[matches[1]] = struct{}{}
+	}
+	sessionMatches := make(map[string]struct{})
+	for _, matches := range sessionPat.FindAllStringSubmatch(dataStr, -1) {
+		sessionMatches[matches[1]] = struct{}{}
+	}
+
+	// Process matches.
+	for idMatch := range idMatches {
+		if detectors.StringShannonEntropy(idMatch) < aws.RequiredIdEntropy {
+			continue
+		}
+		if s.skipIDs != nil {
+			if _, ok := s.skipIDs[idMatch]; ok {
+				continue
+			}
+		}
+
+		for secretMatch := range secretMatches {
+			if detectors.StringShannonEntropy(secretMatch) < aws.RequiredSecretEntropy {
+				continue
+			}
+
+			for sessionMatch := range sessionMatches {
+				if detectors.StringShannonEntropy(sessionMatch) < 4.5 {
```

Comment:
question: Should this be comparing against the var/const `RequiredSecretEntropy`?

Reply:
No. The secret key is a shorter value:
```
os.environ['AWS_SECRET_ACCESS_KEY'] = 'Rtkz4cCIUCp+L9ubINheX6ms/e2wV2ErksMFHzuQ'
```

The session key is much longer:
```
os.environ['AWS_SESSION_TOKEN'] = 'FwoGZXIvYXdzEEUaDLkfiW1wPZNAXd6WESLXAZb4AL9sF87W9TGDZRf8WFb42sb3c1vMl6yNa+4vFZ9uTrNJvO8vlGpoGPsLAShXn2fSIM63r3uuIpApLnGwqVDPoCn4lH6QPmfJgT26AKYP7mA9lerKuycosON5Cx14z67Hp9SfqbbTo8N2vEjui93FYNGsUM/vp/hbWljlSxLJHfG5uML9Q7kb907blfMrF/g9Yh2V/AgJn66qyVJ/iIqzMcaWjbLmJA54i1sx6Px9OiANswhEU+thL+rU2E2lqV69uLvRii53573ap4T/3H+fmwUdNalVKNDygqgGMi0f+1SXMob1qP2H9q1Lpmn46H6s8g18FkOpTuFi1TLDPjeWuogdFRmX+E8jc2I='
```